editor/src/scripts/fetch-and-parse-tailwind-classes.ts

```diff
@@ -0,0 +1,110 @@
+/* eslint-disable no-console */
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+import { mapValues } from '../core/shared/object-utils'
+
+const ClassMatcher = new RegExp(String.raw`(?<![\w\-:])\.[^\}]+\}{1}`, 'g')
+const CustomPseudoClassMatcher = new RegExp(String.raw`(.+)\\(:.+):\1(.+)`)
+const IgnoringCombinatorsMatcher = new RegExp(String.raw`([^>\|\~\+\s]+).*\{(.+)\}`)
+
+function fixCustomPseudoClass(rule: string): string {
+  // Tailwind defines custom pseudoclasses by prepending and appending and using '\:' e.g. focus\:sr-only:focus
+  const matches = CustomPseudoClassMatcher.exec(rule)
+  if (matches == null) {
+    return rule
+  } else {
+    return `${matches[1]}${matches[2]}${matches[3]}`
+  }
+}
+
+function separateClassNameFromBody(
+  rule: string,
+): {
+  className: string
+  body: string
+} {
+  // Some of the class selectors are defined using combinators, in which case we're only interested in the class name
+  const matches = IgnoringCombinatorsMatcher.exec(rule)
+  if (matches == null) {
+    throw new Error(`Failed to separate rule into className and body for ${rule}`)
+  } else {
+    const unadjustedClassName = matches[1]
+    const className = unadjustedClassName.replace(/\\/g, '')
+
+    return {
+      className: className,
+      body: matches[2],
+    }
+  }
+}
+
+function extractAttributeKeys(ruleBody: string): Array<string> {
+  const lines = ruleBody.split(`;`)
+  return lines.map((line) => line.split(':')[0])
+}
+
+type ClassNameToAttributes = { [className: string]: Array<string> }
+
+function parseClasses(rawCSS: string): ClassNameToAttributes {
+  const matches = rawCSS.match(ClassMatcher)
+  let parsedClasses: { [className: string]: Set<string> } = {}
+
+  matches?.forEach((match) => {
+    try {
+      const withoutDot = match.slice(1) // Remove the dot prefix
+      const withFixedPseudoClasses = fixCustomPseudoClass(withoutDot)
+      const { className, body } = separateClassNameFromBody(withFixedPseudoClasses)
+      const attributes = extractAttributeKeys(body)
+
+      let existing = parsedClasses[className] ?? new Set()
+      attributes.forEach((a) => existing.add(a))
+      parsedClasses[className] = existing
+    } catch (e) {
+      throw new Error(`Failed on match ${match}: ${e.message}`)
+    }
+  })
+
+  return mapValues((set: Set<string>) => Array.from(set), parsedClasses)
+}
+
+async function fetchFullCSS(): Promise<string> {
+  const response = await fetch('https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css', {
+    method: 'GET',
+  })
+  if (response.ok) {
+    return response.text()
+  } else {
+    throw new Error(
+      `Failed to fetch latest tailwind css (${response.status}): ${response.statusText}`,
+    )
+  }
+}
+
+const SmolSample = [
+  String.raw`/*! tailwindcss v2.2.4 | MIT License | https://tailwindcss.com *//*! modern-normalize v1.1.0 | MIT License | https://github.com/sindresorhus/modern-normalize */`,
```

Comment:
oh I didn't know about `String.raw` but this could have been useful when making the projectBundle scripts!

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/raw

If I understand correctly: String.raw parses the string with the escape character behavior (for example `\n` makes a newline) switched off. So
```
String.raw`hi\nthere` 
```
will parse `\` and `n` as two regular characters: `['h', 'i', '\', 'n', 't'...]`

Reply:
Correct. We're using it elsewhere with Regexs, because otherwise there is a load of confusion when trying to use regex escaping, especially when using an external tool (e.g. https://regexr.com/) to craft a regex and then trying to correctly escape that when pasting it into your code.